### ISSUE TITLE
getContent() handles web maps/scenes

### ISSUE THREAD

**Issue by tomwayson** [Esri/hub.js#401]
Related to #322, when `getContent()` fetches a web map, it should return the additional information required to render a web map.

The pseudo logic from the Hub app's `composeWebMap()` is 
```
if (isHubAuthoredWebmap(model)) {
  if (!model.orgId) {
    // the org id is needed because so we can determine if we are viewing content that was created
    // by a community member of _this_ Hub (by comparing the author `orgId` and `communityOrgId`)
    model.orgId = await getUserByName(model.owner).orgId
  }
  // fetch the item's data and hang it off of itemData
  model.itemData = await getDataById(model.id);
  // create an item property that effectively either:
  // - in the case of items points to the model itself (blowing up JSON.stringify/parse)
  // - in the case of actual dataset models is a shallow copy of all attributes (even if they aren't part of the item spec)
  model.item = model.attirbutes ? shallowCopy(model.attributes) : model
}
```

I have a few questions about how to proceed here which I'll put in individual comments below.

**Comment by tomwayson** [comment]
The first question is whether or not we should leave the bit about forcing the `orgId` for hub authored web maps. @JoshCrozier and @jPurush do you know any more specifics how that `orgId` is used in the app? Do we show a list of the web maps authored by members of _this Hub's_ community? Or is it used to gate editing? Unless that use case(s) would apply outside the app in say something like a Stencil component (`<hub-web-map id="awef" hub-community-org-id="fawefa" />`???), I think we should leave that in the app for now (just apply it in the content views as we as composer).

**Comment by tomwayson** [comment]
The next question is how to handle the `item` and `data` properties. In hub.js we have many entities that extend from `IModel`:

https://github.com/Esri/hub.js/blob/ec4c1602b9eee05d1a9e37e23fc8f6df809c81ae/packages/common/src/types.ts#L15-L22

So we could have `IHubContent` extend `IModel` and add `item` and `data` properties. @drewctate and I discussed this when talking about how to have `getContent()` better support solution and template content views.

I think if we do nothing else we will fetch the item's data and expose it via `content.data`. For historical reasons (ease of use w/ JSAPI 3.x) we have used the name `itemData` instead of `data` for web maps, but I think going forward it would be better to follow suit w/ other item types and use `data` and update the Hub app to expect that instead of `itemData` (cc/ @JoshCrozier and @jPurush).

**Comment by tomwayson** [comment]
Lastly, I'm less certain of what to do about the item info, b/c we've also that `IHubContent` extends `IItem`, so all the item information is already baked in as a subset of the top level properties of `this`. We don't want to do the lazy `content.item = content` which breaks reading/writing as JSON. I'm also hesitant to make a copy of the item info, but that may be what we need to do unless we want to start using a class like `const content = new Content(datasetModelOrItem)` where the source data is stored in `content.item` and we use top level getters for all the duplicate or derived properties. That's starting to sound more and more appealing, but up until now we've decided that our data fetching functions should return POJOs and not instances.

Should we copy the item info into `content.item`? That's certainly the easiest path forward b/c it allows our existing app code to work and provides an easy path for other types of content (like solutions and templates) to start expecting an `IHubContent` instead of just an `IModel`. It feels dirty though.

**Comment by tomwayson** [comment]
Thanks @JoshCrozier - to me that means we keep that logic in the app, at least for now.

**Comment by tomwayson** [comment]
If we decide to stick w/ POJOs and copy the item info to `content.item`, I would suggest that at the next breaking change we drop the duplicate properties at the top level.

**Comment by ajturner** [comment]
@tomwayson @JoshCrozier We want to display "Community Created" label on any content. Currently we only allow for community members to create webmap in Hub, but they can create other content in AGO that would be shared through a Site. We have a high-priority backlog to enable community members to create other content, Teams, Events, etc. So showing the original org and "Official" vs. "Community" is a comprehensive need. To your original question, this also should apply for embeds outside of Hub in stencil. So if I an org embeds a Gallery in their wordpress site of "community suggestions" it would have this label as well.

So I recommend that we pull `orgId` into the `IHubResource` or `IContent` since all things in Hub would benefit from showing it's source organization.

**Comment by ajturner** [comment]
@tomwayson  to your question on POJO vs. Class instances. I agree we want to be pragmatic with migrating Hub UI to use Hub models. Meaning Hub UI doesn't specifically know about Item or Data underlying details. 

Lazy load getters via a Class does seems the most elegant solution. This works for current implementation `content.item.title` as well as `content.title` which may have to fetch Item first from Portal API. This is future leaning that should prevent a refactor if/when Hub API has all content (public/private) and again removes the underlying implementation detail of where a particular property is stored. That will be apparent with `content.boundary` which may fetch from item/resources, or `content.templateParent` which may be item.properties, or item.typekeywords, or item.data.parentId.

**Comment by drewctate** [comment]
@tomwayson @ajturner I agree in principle with the elegance of a simple class that uses the `item` and `data` POJOs basically as data transfer objects that feed data into a standardized top-level API of getters. That said, I'm not sure what "lazy load getters" would entail. If they entail what I think they do then every getter would have to return a promise forcing a client to do things like
```
const title = await content.title;
```
which doesn't feel very ergonomic to me. In that scenario, it's also hard for the client to know the performance ramifications of accessing various content properties (i.e. what needs to be loaded for this property to be filled and how long will it take).

It's worth noting that certain pieces of the application, especially around templating, do expect an `IModel`, so if we go with a `Content` class, there needs to be a way to get an `IModel`-looking thing out of it. But that wouldn't be too hard. Something like `const model = content.model()` perhaps.

**Comment by tomwayson** [comment]
Thanks for your replies @ajturner  and @drewctate.

Let's take the notion of lazily loaded getters out of this for now. Not only does it up the complexity by an order of magnitude, it's not a widely used pattern (if not an anit-pattern).

@drewctate now that #402 lands you can easily get a "model" out of (at least map) content like:

```js
const model = {
  item: content,
  data: content.data
}
```

If we go so far as to create a class we could indeed add a `.model()` method that returns that^^^, but I think it would be better to just implement `IModel` (i.e. it has `.item` and `.data` properties). 

Based on this discussion I'm going to give that a go and see how it feels before just stabbing duplicate copy of the item info on `content.item`.

**Comment by tomwayson** [comment]
I looked into creating a class, but there were more questions than answers. Mostly b/c I think we're doing TS wrong w/r/t types, interfaces, and inheritance, but also a bit around how much of the item do we want to expose at the top level (cause right now we just throw the whole enchilada up there). See my inline comments below for more specifics, but I think at least I need to go to TS school and it will be a breaking change before we can do something like this.

```ts
class HubContent implements IHubContent {
  // this can only get set in the constructor
  readonly item: IItem;
  // properties that get their defaults when the item
  // is set in the constructor but can be overridden (mutable)
  // NOTE: there are many others defined on IHubContent, but
  // TS complains if we don't re-define these required ones
  hubId: string;
  publishedDate: Date;
  // pass through item properties
  // NOTE: these are only the required ones so far
  get id() {
    return this.item.id
  }
  get title() {
    return this.item.title
  }
  get owner() {
    return this.item.owner
  }
  get tags() {
    return this.item.tags
  }
  get created() {
    return this.item.created
  }
  get modified() {
    return this.item.modified
  }
  get size() {
    return this.item.size
  }
  get type() {
    return this.item.type
  }
  get numViews() {
    return this.item.numViews
  }
  // TODO: define any remaining ones we use, but
  // we probably don't want to define all of these:
  // typeKeywords,
  // description,
  // snippet,
  // thumbnail,
  // extent,
  // categories,
  // spatialReference,
  // accessInformation,
  // licenseInfo,
  // culture,
  // url,
  // access,
  // orgId
  // proxyFilter,
  // properties,
  // appCategories,
  // industries,
  // languages,
  // largeThumbnail,
  // banner,
  // screenshots,
  // listed,
  // ownerFolder,
  // size,
  // protected
  // commentsEnabled,
  // numComments,
  // numRatings,
  // avgRating,
  // itemControl,
  // scoreCompleteness,
  // orgId,
  // add getters for item properties that we squash
  // TODO: maybe these aren't needed since we can just access the item now
  get itemName() {
    return this.item.name
  }
  get itemCategories() {
    return this.item.categories
  }
  // properties derived from item
  get name() {
    return this.item.title
  }
  get createdDate() {
    return new Date(this.item.created)
  }
  get updatedDate() {
    return new Date(this.item.modified)
  }
  get hubType() {
    return getItemHubType(this.item)
  }
  get permissions() {
    return {
      visibility: this.item.access,
      control: this.item.itemControl || "view"
    }
  }
  // constructors
  constructor(item: IItem) {
    // hold on to the item
    this.item = item

    // set default values of mutable properties based on item
    this.hubId = this.id;
    // TODO: why do we need to cast as IHubContent
    (this as IHubContent).normalizedType = normalizeItemType(item)
    // TODO: others... which also need this to be cast to IHubContent, why???
    // this.publishedDateSource
    // this.actionLinks
    // this.hubActions
    // this.metrics
    // this.visibility
    // this.slug
    // this.portalDataUrl
    // this.groupIds
    // this.errors = [];
    // TODO: categories should be flattened like they are in dataset
    // this.categories = flatten(item.categories)
    this.publishedDate = this.createdDate
    // TODO: remaining defaults
    // TODO: why do these fail w/ errors like
    // Property 'summary' does not exist on type 'HubContent'.
    // b/c they are all on IHubResource, would we need to cast to that?
    // this.summary = item.snippet || item.description
    // this.createdDateSource = 'item.created'
    // this.updatedDateSource = 'item.modified'
    // this.boundary = ...
  }
}
```

I think for now I'mma just stab a copy of the item on there and call it a day.
